### PR TITLE
RIA-8157 update tribunal decision

### PR DESCRIPTION
**Pull request by jamieclarkeHMCTS**
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/RIA-8157

Change description
Update Tribunal decision - Event and screens for correct decision under R31

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionRule31MidEventTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.Event;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+
+import java.util.Optional;
+
+@MockitoSettings(strictness = Strictness.LENIENT)
+@ExtendWith(MockitoExtension.class)
+@SuppressWarnings("unchecked")
+class UpdateTribunalDecisionRule31MidEventTest {
```
Comment (ThomasKKC) [inline src/test/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionRule31MidEventTest.java:31]: Need to finish the tests for this class

File: src/test/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEventTest.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.field.Document;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.field.YesOrNo;
+
+@MockitoSettings(strictness = Strictness.LENIENT)
+@ExtendWith(MockitoExtension.class)
+@SuppressWarnings("unchecked")
+class UpdateTribunalDecisionErrorRule31MidEventTest {
```
Comment (ThomasKKC) [inline src/test/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEventTest.java:32]: Can you add the "Cannot handle callback" tests here, see the class below.

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -164,6 +164,8 @@ homeOfficeApi:
   aboutToStartPath: "/asylum/ccdAboutToStart"
   aboutToSubmitPath: "/asylum/ccdAboutToSubmit"
 
+decisionAndReasonsAmendedPdf.fileName: "Decision-and-reasons"
```
Comment (ThomasKKC) [inline src/main/resources/application.yaml:167]: Is this used anywhere? We can remove it.

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/postsubmit/UpdateTribunalDecisionConfirmation.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    ) {
+        if (!canHandle(callback)) {
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        PostSubmitCallbackResponse postSubmitResponse =
+            new PostSubmitCallbackResponse();
+
+        postSubmitResponse.setConfirmationHeader("# You have updated the decision");
+        postSubmitResponse.setConfirmationBody(
```
Comment (ThomasKKC) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/postsubmit/UpdateTribunalDecisionConfirmation.java:34]: Please add confirmation functional test.

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.Event;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.field.Document;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.field.YesOrNo;
+import uk.gov.hmcts.reform.iacaseapi.domain.handlers.PreSubmitCallbackHandler;
+
+@Component
+public class UpdateTribunalDecisionErrorRule31MidEvent implements PreSubmitCallbackHandler<AsylumCase> {
```
Comment (ThomasKKC) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java:23]: Can we rename this class? I know previously it was for error only, but we have merged the document name update with this class.

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        final AsylumCase asylumCase = callback.getCaseDetails().getCaseData();
+        PreSubmitCallbackResponse<AsylumCase> response = new PreSubmitCallbackResponse<>(asylumCase);
+
+        final Optional<Document> decisionAndReasonsDoc = asylumCase.read(DECISION_AND_REASON_DOCS_UPLOAD, Document.class);
+
+        YesOrNo isDecisionAndReasonDocumentBeingUpdated = asylumCase.read(AsylumCaseFieldDefinition.UPDATE_TRIBUNAL_DECISION_AND_REASONS_FINAL_CHECK, YesOrNo.class)
+            .orElse(NO);
+        Optional<DynamicList> optionalTypesOfUpdateTribunalDecision = asylumCase.read(AsylumCaseFieldDefinition.TYPES_OF_UPDATE_TRIBUNAL_DECISION, DynamicList.class);
+
+        if (isDecisionAndReasonDocumentBeingUpdated == YES) {
```
Comment (ThomasKKC) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java:50]: Can we use .equals(YES) here? YesOrNO id is a String

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            }
+
+            Document previousDecisionAndReasonsDoc = decisionAndReasonsDoc.get();
+
+            Document amendedDecisionAndReasonsDoc = new Document(previousDecisionAndReasonsDoc.getDocumentUrl(), previousDecisionAndReasonsDoc.getDocumentBinaryUrl(),getDecisionAndReasonsFilename(asylumCase));
+
+            asylumCase.write(DECISION_AND_REASON_DOCS_UPLOAD, amendedDecisionAndReasonsDoc);
+            return new PreSubmitCallbackResponse<>(asylumCase);
+
+        } else if (isDecisionAndReasonDocumentBeingUpdated == NO &&
```
Comment (ThomasKKC) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java:69]: Same here, can we use .equals(No)?

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        final Optional<Document> decisionAndReasonsDoc = asylumCase.read(DECISION_AND_REASON_DOCS_UPLOAD, Document.class);
+
+        YesOrNo isDecisionAndReasonDocumentBeingUpdated = asylumCase.read(AsylumCaseFieldDefinition.UPDATE_TRIBUNAL_DECISION_AND_REASONS_FINAL_CHECK, YesOrNo.class)
+            .orElse(NO);
+        Optional<DynamicList> optionalTypesOfUpdateTribunalDecision = asylumCase.read(AsylumCaseFieldDefinition.TYPES_OF_UPDATE_TRIBUNAL_DECISION, DynamicList.class);
+
+        if (isDecisionAndReasonDocumentBeingUpdated == YES) {
+
+            if (decisionAndReasonsDoc.isEmpty()) {
+                response.addError("Amended Decision And Reasons Document must be present");
```
Comment (ThomasKKC) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java:53]: Do we need this? It is not part of the ticket criteria, the "MANDATORY" on CCD def will catch this

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            Document previousDecisionAndReasonsDoc = decisionAndReasonsDoc.get();
+
+            Document amendedDecisionAndReasonsDoc = new Document(previousDecisionAndReasonsDoc.getDocumentUrl(), previousDecisionAndReasonsDoc.getDocumentBinaryUrl(),getDecisionAndReasonsFilename(asylumCase));
+
+            asylumCase.write(DECISION_AND_REASON_DOCS_UPLOAD, amendedDecisionAndReasonsDoc);
+            return new PreSubmitCallbackResponse<>(asylumCase);
+
+        } else if (isDecisionAndReasonDocumentBeingUpdated == NO &&
+            optionalTypesOfUpdateTribunalDecision.isPresent() &&
+            "No".equals(optionalTypesOfUpdateTribunalDecision.get().getValue().getLabel())) {
```
Comment (ThomasKKC) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java:71]: For visibility, can we flip the equals on this?
(optionalTypesOfUpdateTribunalDecision.get().getValue().getLabel()..equals("No")

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            response.addError("You must update the decision or the Decision and Reasons document to continue");
```
Comment (ThomasKKC) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionErrorRule31MidEvent.java:72]: You might need to add full stop here, the prototype show it, can check with Elli

**Review by ThomasKKC — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionRule31MidEvent.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        DynamicList typesOfUpdateTribunalDecision = new DynamicList(new Value("", ""), values);
+
+        asylumCase.write(TYPES_OF_UPDATE_TRIBUNAL_DECISION, typesOfUpdateTribunalDecision);
+        return new PreSubmitCallbackResponse<>(asylumCase);
+    }
+
+    private boolean isDecisionAllowed(AsylumCase asylumCase) {
+        return asylumCase
+            .read(IS_DECISION_ALLOWED, AppealDecision.class)
+            .map(type -> type == ALLOWED).orElse(false);
```
Comment (ikirsanov) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/UpdateTribunalDecisionRule31MidEvent.java:71]: equals (as above Thomas mentioned)

**Review by ikirsanov — APPROVED** [review]

**Review by ThomasKKC — DISMISSED** [review]

**Review by ThomasKKC — APPROVED** [review]

**Review by DennisHoV1 — APPROVED** [review]

**Review by MiguelMartinRojas — APPROVED** [review]

**Pull request merged**